### PR TITLE
[Prototype] Multimodal (vision) support

### PR DESCRIPTION
**Pull request by sohamparikh**
# ✨ Description

Multi-modal support, starting with pixtral's vision encoder

## 🔍 Type of change

Select all that apply:

- [ ] 🐛 **Bug fix** (non-breaking change that addresses a specific issue)
- [ ] 🚀 **New feature** (non-breaking change that adds functionality)
- [ ] ⚠️ **Breaking change** (a change that could affect existing functionality)
- [ ] 📈 **Performance improvement/optimization** (improves speed, memory usage, or efficiency)
- [ ] 🛠️ **Code refactor** (non-functional changes that improve code readability, structure, etc.)
- [ ] 📦 **Dependency bump** (updates dependencies, including Dockerfile or package changes)
- [ ] 📝 **Documentation change** (updates documentation, including new content or typo fixes)
- [ ] 🔧 **Infrastructure/Build change** (affects build process, CI/CD, or dependencies)

## 📝 Changes

List the key changes introduced in this PR:

1. `prepare` now optionally takes images and image positions (where they should appear in the text). Images are stored in the memmap file along with text
2. `GPTBaseModel` optionally has a vision encoder attached (can have audio/video encoders in future), which consists of a conv2D, vision transformer and a MLP adapter.

### CODE REVIEW

File: fast_llm/data/preparator/gpt_memmap/prepare.py [hunk 1]
```diff
@@ -321,6 +324,12 @@ def run(self) -> None:
 
         # Calculate total number of tokens
         total_tokens = sum(tqdm.tqdm(tokenized_dataset["num_tokens"], desc="Counting tokens", unit="tokens"))
+        total_pixels = (
+            sum(tqdm.tqdm(tokenized_dataset["num_pixels"], desc="Counting pixels", unit="pixels"))
+            if self._config.dataset.images
+            else 0
+        )
+        total_tokens += total_pixels // np.dtype(self._data_type.numpy).itemsize
```
Comment (RaymondLi0) [inline fast_llm/data/preparator/gpt_memmap/prepare.py:332]: Should divide by `(image_patch_size**2)` instead?

File: fast_llm/data/dataset/gpt/memmap.py [hunk 2]
```diff
@@ -120,16 +137,52 @@ def _init(self, name: str, prefix: pathlib.Path | str, num_documents: int | None
                         offset=rejected_span_offset + idx * 2 * np.dtype(np.int32).itemsize,
                     )
                 )
+            offset += np.array(self._chosen_spans).nbytes + np.array(self._rejected_spans).nbytes
+
+        self._num_pixels = 0
+        self._image_lengths = None
```
Comment (RaymondLi0) [inline fast_llm/data/dataset/gpt/memmap.py:143]: maybe rename to `_image_sizes` ?

File: fast_llm/data/dataset/gpt/sampled.py [hunk 3]
```diff
@@ -467,27 +531,98 @@ def __getitem__(self, index: int) -> typing.Any:
                     else:
                         # Move on to the next sample.
                         token_count += padding_size
+                        continue
+                elif document_size + tokens_in_sample == self._parameters.sequence_length + 1:
+                    if token_count + document_size == token_start:
+                        token_count += document_size
+                        document_sampling_index += 1
+                        continue
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/data/dataset/gpt/sampled.py:539]: yes, i have a separate PR for that. Added it here to unblock training
https://github.com/ServiceNow/Fast-LLM/pull/287

File: fast_llm/data/preparator/gpt_memmap/prepare.py (hunk 1 above)
Comment (sohamparikh) [reply fast_llm/data/preparator/gpt_memmap/prepare.py:332]: This is something I couldn't cleanly solve. `total_tokens` here is used to determine the shard size, so I got the token-equivalent bytes for pixels. Everywhere else we use `image_patch_size**2` to determine the actual number of image tokens and weigh the shard accordingly for blending

File: fast_llm/layers/transformer/vision_transformer.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+from fast_llm.engine.config_utils.tensor_space import TensorDim
+from fast_llm.layers.transformer.config import VisionTransformerKwargs
+from fast_llm.layers.transformer.transformer import TransformerLayer
+from fast_llm.tensor import TensorMeta
+
+
+class VisionTransformerLayer(TransformerLayer):
+    _name: str = "Vision transformer layer"
+
+    def _get_meta(self, tensor: torch.Tensor, name: str, kwargs: dict):
```
Comment (RaymondLi0) [inline fast_llm/layers/transformer/vision_transformer.py:12]: This seems to be the same as in the base class?
If it's the case do we need this class at all?

File: fast_llm/data/preparator/gpt_memmap/prepare.py [hunk 5]
```diff
@@ -298,11 +305,7 @@ def run(self) -> None:
             raise ValueError(f"Both chosen and rejected loss masking spans must be specified if one is specified.")
 
         # route tokenize function
-        if self._config.dataset.loss_masking_spans is not None:
-            if self._config.dataset.loss_masking_spans not in dataset.column_names:
-                raise ValueError(f"Dataset does not have spans field '{self._config.dataset.loss_masking_spans}'.")
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/data/preparator/gpt_memmap/prepare.py:303]: i moved everything to a single tokenize function, since the combinations were getting too much (images, loss spans, audio next)

File: fast_llm/data/preparator/gpt_memmap/prepare.py (hunk 1 above)
Comment (RaymondLi0) [reply fast_llm/data/preparator/gpt_memmap/prepare.py:332]: Oh I see, that makes sense yes! Added a comment as a suggestion
```suggestion
        # Add the token-equivalent bytes of pixels to determine shard size
        total_tokens += total_pixels // np.dtype(self._data_type.numpy).itemsize
```

File: fast_llm/layers/vision_encoder/config.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    patch_size = "vision_patch_size"
+    kv_channels = "vision_kv_channels"
+
+
+class VisionEncoderKwargs:
+    patch_size = "patch_size"
+    images = "images"
+    image_patches = "image_patches"
+    image_positions = "image_positions"
+    image_size = "image_size"
```
Comment (RaymondLi0) [inline fast_llm/layers/vision_encoder/config.py:25]: Maybe rename to `max_image_size` to avoid confusion with `image_sizes`? (rename in `BatchConfig` as well if we rename here)

File: fast_llm/data/tokenizer.py [hunk 7]
```diff
@@ -35,44 +35,84 @@ def vocab(self) -> dict[str, int]:
...
         char_pos = 0
+        token_ids = []
+        image_token_positions = []
+        token_spans = []
         beginning_of_text = True
-        
+        image_position = image_positions[image_idx] if image_idx < len(image_positions) else float("inf")
+
         for start, end in char_spans:
+            while image_position <= start:
```
[the remark this answers is not in the document]
Comment (sohamparikh) [reply fast_llm/data/tokenizer.py]: does it simplify things now?

File: fast_llm/data/dataset/gpt/sampled.py [hunk 8]
```diff
@@ -92,6 +95,11 @@ def __init__(
         self._truncate_documents = sampling.parameters.truncate_documents
         self._device = torch.device("cuda" if self._config.gpu else "cpu")
 
+        if self._indexed_dataset.has_images and self._truncate_documents:
+            raise RuntimeError(
+                "Truncating documents with images is not yet supported. Please turn off truncation to use images."
+            )
```
Comment (tscholak) [inline fast_llm/data/dataset/gpt/sampled.py:101]: what does that mean in practice? documents with images that are longer than the sequence length are discarded?

File: fast_llm/data/dataset/gpt/sampled.py [hunk 9]
```diff
@@ -133,24 +141,48 @@ def _sample(self) -> None:
...
 
         # Calculate basic stats.
         if not self._truncate_documents:
             assert _extension_available, (
                 "The C++ extension for dataset sampling is missing."
                 " Please make sure Fast-LLM is installed correctly."
             )
-            long_docs_filter = document_sizes > self._parameters.sequence_length + 1
+            long_docs_filter = document_sizes + image_token_sizes > self._parameters.sequence_length + 1
             ignored_documents = long_docs_filter.sum().item()
```
Comment (tscholak) [inline fast_llm/data/dataset/gpt/sampled.py:179]: I guess yes, long docs with images will be ignored. ok

File: fast_llm/models/gpt/config.py [hunk 10]
```diff
@@ -71,6 +71,17 @@ class DiffusionLlamaGPTHuggingfaceCheckpointFormat(GPTHuggingfaceCheckpointForma
     trust_remote_code: typing.ClassVar[bool] = True
 
 
+class LlavaGPTHuggingfaceCheckpointFormat(GPTHuggingfaceCheckpointFormat):
```
Comment (jlamypoirier) [inline fast_llm/models/gpt/config.py:74]: I don't understand, is all that added complexity just so we can call it `llava` instead of either `pixtral` or `mistral`?

File: fast_llm/models/gpt/model.py [hunk 11]
```diff
@@ -63,6 +70,10 @@ def __init__(
         if self._config.enable_dpo:  # TODO better way to pass in?
             self._preprocessors.append(PreferenceSpanPreprocessor(self._config, self._tensor_space))
 
+        if self._config.vision_encoder.enabled:
```
Comment (jlamypoirier) [inline fast_llm/models/gpt/model.py:73]: I think this would be more appropriate as a separate model, like we did for SSM

**Review by jlamypoirier — COMMENTED** [review]
Looks good, but I'm worried about the added complexity. Main suggestions:
* Make vision into a separate model.
* Replace the dim and kwarg name changes with simpler alternatives.
* Break down some methods that have grown too big to be properly understandable.
* Avoid abbreviations when possible so names are self-explanatory for everyone.

File: fast_llm/models/gpt/model.py (hunk 11 above)
Comment (tscholak) [reply fast_llm/models/gpt/model.py:73]: No, this is intentionally part of the base GPT model. It ensures that multimodal support is transparent and inherited by all model variants (including SSM) without needing separate parallel class hierarchies. We're not introducing architectural silos unless there's a concrete need.

File: fast_llm/models/gpt/conversion.py [hunk 12]
```diff
@@ -544,10 +592,10 @@ def _get_mlp_converters(self, fast_llm_prefix: str, hf_prefix: str) -> list[Weig
 
 class MistralHuggingfaceCheckpointHandler(CommonLlamaHuggingfaceCheckpointHandler):
     format: typing.ClassVar[type[CheckpointFormat]] = MistralGPTHuggingfaceCheckpointFormat
-    architecture: typing.ClassVar[str] = "MistralForCausalLM"
 
     @classmethod
     def _create_config_converters(cls) -> list[ParamConverter]:
+        cls.architecture = "MistralForCausalLM"
```
Comment (RaymondLi0) [inline fast_llm/models/gpt/conversion.py:598]: Why are the `architecture` classvars modified here?

File: fast_llm/engine/multi_stage/stage.py [hunk 13]
```diff
@@ -137,7 +137,7 @@ def backward(
         assert self._mode.support_backward
         input_, output = grad_context
         output.backward(output_grad)
-        return input_.grad
+        return input_.grad if input_.grad is not None else torch.zeros_like(input_)
```
Comment (jlamypoirier) [inline fast_llm/engine/multi_stage/stage.py:140]: That's not a good idea, it will cause unnecessary operations when not needed, which is most of the time.

File: fast_llm/functional/config.py [hunk 14]
```diff
@@ -67,7 +68,8 @@ def _set_activation_fn_map() -> None:
     global _ACTIVATION_FN_MAP
 
     _ACTIVATION_FN_MAP = {
-        ActivationType.gelu: lambda x: torch.nn.functional.gelu(x, approximate="tanh"),
+        ActivationType.gelu: torch.nn.functional.gelu,
```
Comment (jlamypoirier) [inline fast_llm/functional/config.py:71]: This is incorrect, Fast-LLM uses the tanh version.

File: fast_llm/data/dataset/gpt/indexed.py [hunk 15]
```diff
@@ -19,12 +19,26 @@ def get_document_sizes(self) -> np.ndarray:
         and derived classes should try to avoid holding the whole array im memory.
         """
 
+    @abc.abstractmethod
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/indexed.py:22]: I separated this because there were several issues with the previous version.

File: fast_llm/data/dataset/gpt/memmap.py [hunk 16]
```diff
@@ -1,8 +1,10 @@
+import io
```
Comment (jlamypoirier) [inline fast_llm/data/dataset/gpt/memmap.py:1]: Reworked this file so the various concepts are properly compartmentalized. Next step would be to extract those components into separate classes, it's not strictly needed yet but will become if we keep adding stuff.

File: fast_llm/models/gpt/model.py [hunk 17]
```diff
@@ -161,15 +227,28 @@ def preprocess_meta(
...
+                }
+            )
 
         common_kwargs = {
             LanguageModelKwargs.phase: phase,
             TransformerKwargs.sequence_first: sequence_first,
             TransformerKwargs.hidden_dims: hidden_dims,
             TransformerKwargs.sequence_length: sequence_length,
             TransformerKwargs.sequence_q_dim: sequence_q_dim,
-            LanguageModelKwargs.mask_inputs: not truncate_documents,
```
[the remark this answers is not in the document]
Comment (RaymondLi0) [reply fast_llm/models/gpt/model.py:171]: I think this was unintentionally removed in a merge from main
